### PR TITLE
fix(otel): map GenAI IO on AI SDK spans

### PR DESCRIPTION
**Pull request by HarperZ9**
## Summary

- Map official `gen_ai.input.messages` / `gen_ai.output.messages` on AI SDK scope generation spans when `ai.*` I/O fields are absent.
- Map official `gen_ai.tool.call.arguments` / `gen_ai.tool.call.result` on AI SDK scope tool spans when `ai.*` I/O fields are absent.
- Add an OTel mapping regression test covering AI SDK scope spans that emit official GenAI I/O attributes.

## Context

This addresses the mapping gap reported in #14633: AI SDK 7 can emit spans under instrumentation scope `ai` while carrying official `gen_ai.*` I/O attributes. The existing AI SDK branch returned before the generic OpenTelemetry I/O mapper, so those attributes were removed from metadata but not promoted to observation input/output.

The fix keeps existing `ai.*` fields as the priority path and only falls back to official `gen_ai.*` fields when the AI SDK-specific fields are missing.

## Verification

- `pnpm exec prettier --check packages/shared/src/server/otel/OtelIngestionProcessor.ts web/src/__tests__/server/api/otel/otelMapping.servertest.ts`
- `git diff --check`
- `pnpm --filter @langfuse/shared run db:generate`
- `pnpm --filter @langfuse/shared run typecheck`

I also attempted the targeted Vitest server test locally on Windows, but this checkout's Vitest server project collected zero server files even without a path filter while still printing `src\__tests__\server\api\otel\otelMapping.servertest.ts` in the generated include list. I did not count that as a passing test run.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR closes a mapping gap for AI SDK 7 spans emitted under instrumentation scope `ai` that carry official `gen_ai.*` I/O attributes instead of the `ai.*`-prefixed ones. The fix appends a nullish-assignment fallback at the end of the existing `ai`-scope branch so that `gen_ai.input.messages`/`gen_ai.output.messages` and `gen_ai.tool.call.arguments`/`gen_ai.tool.call.result` are promoted to observation `input`/`output` when no `ai.*` I/O field is present.

- **Processor change** (`OtelIngestionProcessor.ts`): seventeen new lines inside the `instrumentationScopeName === "ai"` guard; uses `??=` to preserve the existing `ai.*` priority chain and only apply the GenAI fallback when those fields are absent. All four `gen_ai.*` keys are already in `potentialInputOutputKeys` (lines 1488–1491), so they are stripped from `filteredAttributes` before the branch runs.
- **Test** (`otelMapping.servertest.ts`): new regression test asserts that values are lifted to `body.input`/`body.output` and absent from `metadata.attributes` for both a generation span and a tool span. One of the four filtered attributes (`gen_ai.input.messages` on the generation span) is not explicitly checked for metadata absence, leaving a small gap in the filter-cleanup assertions.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

Safe to merge; the production logic is correct and the change is tightly scoped to the AI SDK instrumentation-scope branch.

The fallback uses ??= correctly so existing ai.* mappings are never displaced. All four gen_ai.* keys were already in the filter list before this PR, so they will not bleed into metadata. The only gap is a missing test assertion for gen_ai.input.messages being absent from generation-span metadata — a cosmetic omission that does not affect runtime behaviour.

The test file (otelMapping.servertest.ts) is missing one of four expected metadata-filter assertions; the processor file (OtelIngestionProcessor.ts) needs no further attention.
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
web/src/__tests__/server/api/otel/otelMapping.servertest.ts:4682-4684
The test verifies that `gen_ai.output.messages`, `gen_ai.tool.call.arguments`, and `gen_ai.tool.call.result` are stripped from metadata, but the symmetric check for `gen_ai.input.messages` on the generation span is missing. While the attribute IS removed (it lives in `potentialInputOutputKeys` on line 1488 of the processor), the test's own goal of confirming filtered-attribute cleanup is stated incompletely. A future refactor of the filter list could silently reintroduce the key without a failing test.

```suggestion
      expect(
        generation?.body.metadata?.attributes?.["gen_ai.input.messages"],
      ).toBeUndefined();
      expect(
        generation?.body.metadata?.attributes?.["gen_ai.output.messages"],
      ).toBeUndefined();
```

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(otel): map GenAI IO on AI SDK spans"](https://github.com/langfuse/langfuse/commit/72a496534e9e6c886a87e887dbd26c5c27bf8133) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=41246842)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

**Comment by philiparpin** [comment]
Thanks for jumping on this so quickly, @HarperZ9 🙏 — this matches the fallback approach proposed in #14633.

Confirming real-world impact: we hit this migrating a production backend to **AI SDK 7** with the official `@langfuse/vercel-ai-sdk` integration on **Langfuse Cloud**. Spans are emitted on instrumentation scope `"ai"` and carry `gen_ai.input.messages`, `gen_ai.output.messages`, `gen_ai.tool.call.arguments`, and `gen_ai.tool.call.result` as attributes, but today only `input` maps — generation `output` and tool `input`/`output` come back `null` (full before/after evidence, incl. raw span-attribute dumps, in #14633).

Keeping the `ai.*` priority and falling through to the official `gen_ai.*` attributes (plus the regression fixture) looks exactly right to us.

Happy to validate end-to-end against our repro (real `ToolLoopAgent` + tool-call traces) once it's reviewable. Two things currently gate review/merge:
- the PR is still a **draft** — mind marking it ready for review when you're set?
- `license/cla` is pending (CLA not signed yet).

Thanks again for the quick turnaround!

**Comment by HarperZ9** [comment]
Thanks @philiparpin. I marked this ready for review, fixed the commit attribution so `license/cla` is green, and rebased onto current `main` to clear the merge conflict.

After the rebase, `main` already has the production GenAI extraction helper, so this PR is now a focused regression test for AI SDK-scope spans carrying official GenAI I/O attributes (`gen_ai.input.messages`, `gen_ai.output.messages`, and `gen_ai.tool.call.*`). Current head is mergeable and the visible CLA/Snyk checks are green.

File: web/src/__tests__/server/api/otel/otelMapping.servertest.ts [hunk 1]
```diff
@@ -4566,6 +4566,130 @@ describe("OTel Resource Span Mapping", () => {
...
+        '[{"role":"user","content":"hello"}]',
+      );
+      expect(generation?.body.output).toBe(
+        '[{"role":"assistant","content":"hi"}]',
+      );
+      expect(tool?.body.input).toBe('{"location":"Berlin"}');
+      expect(tool?.body.output).toBe('{"temperature":21}');
+      expect(
+        generation?.body.metadata?.attributes?.["gen_ai.output.messages"],
+      ).toBeUndefined();
```
[the remark this answers is not in the document]
Comment (HarperZ9) [reply web/src/__tests__/server/api/otel/otelMapping.servertest.ts:4687]: This is covered in the current PR head `346691d4f0ffba0b15a9363a90285478fa8a0453`: it adds the `gen_ai.input.messages` metadata cleanup assertion immediately before the existing `gen_ai.output.messages` assertion. Thanks for the catch.

**Comment by HarperZ9** [comment]
Closing this as superseded by #14655, which merged the production fix and the AI SDK-scope GenAI I/O regression coverage into `main`.

Thank you @philiparpin for the concrete AI SDK 7 reproduction and the offer to validate it end to end; that evidence helped confirm the exact fallback boundary. There is no additional production change worth carrying in this branch now that the canonical fix has landed.
